addons/udes_stock/models/stock_quant_package.py

```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+from collections import defaultdict
+
+
+class StockQuantPackage(models.Model):
+    _name = "stock.quant.package"
+    _inherit = ["stock.quant.package", "mixin.stock.model"]
+    _description = "Package"
+
+    # Enable create packages
+    MSM_CREATE = True
+
+    def assert_not_reserved(self):
+        """ Ensure that the contents of all packages in the recordset are fully
+            unreserved.
+        """
+        reserved_packages = self.filtered(
+            lambda package: package.get_reserved_quantity() > 0
+        )
+        if reserved_packages:
+            raise ValidationError(
+                _(
+                    "Cannot move packages with reserved contents. "
+                    "Please speak to a team leader to resolve the issue.\n"
+                    "Affected packages: %s"
+                )
+                % (", ".join(reserved_packages.mapped("name")))
+            )
+
+    def get_quantities_by_key(self, get_key=lambda q: q.product_id):
```

Comment:
Suggestion: Feels a bit jarring having a lambda as default argument. Personally I think the following would be neater:

```
def get_quantities_by_key(self, get_key=None):
    if get_key is None:
        get_key = lambda q: q.product_id
```

Might be worth getting a second opinion on that though... or if you prefer how it is now feel free to resolve this one. :D

Reply:
I don't mind seeing a lambda function as default augment but that's me, having said that if you do decide to change it Peter maybe also change the has_same_content() function on line 59 to match for consistency?

Reply:
If it's OK i will leave this one